docs/site/Authentication-component-action.md

```diff
@@ -0,0 +1,252 @@
+
+      debug('%s result -', route.describe(), result);
+      this.send(response, result);
+    } catch (error) {
+      this.reject(context, error);
+    }
+  }
+}
+```
+
+</details>
+
+By default, `authenticate` is **not** part of the sequence of actions, so you
+must create a custom sequence and add the authentication action.
+
+An authentication action `AuthenticateFn` is provided by the
+`AuthenticateActionProvider` class.
+
+`AuthenticateActionProvider` is defined as follows:
+
+```ts
+// ------ CODE THAT EXPLAINS THE MECHANISM ---------
+export class AuthenticateActionProvider implements Provider<AuthenticateFn> {
+  constructor(
+    // The provider is instantiated for Sequence constructor,
+    // at which time we don't have information about the current
+    // route yet. This information is needed to determine
+    // what auth strategy should be used.
+    // To solve this, we are injecting a getter function that will
+    // defer resolution of the strategy until authenticate() action
+    // is executed.
+    @inject.getter(AuthenticationBindings.STRATEGY)
+    readonly getStrategy: Getter<AuthenticationStrategy>,
+    @inject.setter(SecurityBindings.USER)
+    readonly setCurrentUser: Setter<UserProfile>,
+  ) {}
+
+  /**
+   * @returns authenticateFn
+   */
+  value(): AuthenticateFn {
+    return request => this.action(request);
+  }
+
+  /**
+   * The implementation of authenticate() sequence action.
+   * @param request The incoming request provided by the REST layer
+   */
+  async action(request: Request): Promise<UserProfile | undefined> {
+    const strategy = await this.getStrategy();
+    if (!strategy) {
+      // The invoked operation does not require authentication.
+      return undefined;
+    }
+
+    const userProfile = await strategy.authenticate(request);
+    if (!userProfile) {
+      // important to throw a non-protocol-specific error here
+      let error = new Error(
+        `User profile not returned from strategy's authenticate function`,
+      );
+      Object.assign(error, {
+        code: USER_PROFILE_NOT_FOUND,
+      });
+      throw error;
+    }
+
+    this.setCurrentUser(userProfile);
+    return userProfile;
+  }
+}
+```
+
+`AuthenticateActionProvider`'s `value()` function returns a function of type
+`AuthenticateFn`. This function attempts to obtain an authentication strategy
+(resolved by `AuthenticationStrategyProvider` via the
+`AuthenticationBindings.STRATEGY` binding). If **no** authentication strategy
+was specified for this endpoint, the action immediately returns. If an
+authentication strategy **was** specified for this endpoint, its
+`authenticate(request)` function is called. If a user profile is returned, this
+means the user was authenticated successfully, and the user profile is added to
+the request context (via the `SecurityBindings.USER` binding); otherwise an
+error is thrown.
+
+Here is an example of a custom sequence which utilizes the `authentication`
+action.
+
+```ts
```

Comment:
in the todo-jwt example, we are using the default sequence that comes with the scaffolded LB app. Maybe we use that instead? Or explain when we want to create a separate sequence?

Reply:
@dhmlau not really :p the todo-jwt example still uses the custom sequence with "authenticate" injected and invoked.

See [code](https://github.com/strongloop/loopback-next/blob/master/examples/todo-jwt/src/sequence.ts#L34).

Reply:
What i meant is that the sequence.ts is created by default when we're scaffolding the app. In the snippet here seems to guide the users to create a new sequence?

Reply:
@dhmlau I rephrased it to be:

> By default, "authenticate" is **not** part of the sequence of actions, so you
must **modify the default sequence** and add the authentication action.

And

> Here is an example of a **modified** sequence which utilizes the `authentication`
action.